### PR TITLE
parsec: catch section/option conflict in validation

### LINKED ISSUE #4955: Parsec: accidentally specifying sections as settings not handled properly
**Describe the bug**

If you specify `[scheduler]main loop = something` instead of `[scheduler][main loop]`, the workflow validates with no warnings or errors.

```ini
[scheduler]
    main loop = 22  # Eh??
    allow implicit tasks = True
[scheduling]
    [[graph]]
        R1 = coffee
```
```console
$ cylc validate .
Valid for cylc-8.0rc4.dev
```

Or for the `[meta]` section, you get traceback:
```ini
meta = 22  # Eh??
[scheduler]
    allow implicit tasks = True
[scheduling]
    [[graph]]
        R1 = coffee
```
```console
$ cylc validate .
Traceback (most recent call last):
  File "~/.conda/envs/cylc8/bin/cylc", line 33, in <module>
    sys.exit(load_entry_point('cylc-flow', 'console_scripts', 'cylc')())
  File "~/github/cylc-flow/cylc/flow/scripts/cylc.py", line 675, in main
    execute_cmd(command, *cmd_args)
  File "~/github/cylc-flow/cylc/flow/scripts/cylc.py", line 283, in execute_cmd
    entry_point.resolve()(*args)
  File "~/github/cylc-flow/cylc/flow/terminal.py", line 226, in wrapper
    wrapped_function(*wrapped_args, **wrapped_kwargs)
  File "~/github/cylc-flow/cylc/flow/scripts/validate.py", line 116, in main
    cfg = WorkflowConfig(
  File "~/github/cylc-flow/cylc/flow/config.py", line 477, in __init__
    self._check_task_event_handlers()
  File "~/github/cylc-flow/cylc/flow/config.py", line 1488, in _check_task_event_handlers
    get_event_handler_data(taskdef.rtconfig, self.cfg)
  File "~/github/cylc-flow/cylc/flow/task_events_mgr.py", line 317, in get_event_handler_data
    for key, value in workflow_cfg['meta'].items():
AttributeError: 'str' object has no attribute 'items'
```

**Expected behaviour**

An `IllegalItemError` should be raised.

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

### PR DESCRIPTION
**Pull request by oliver-sanders**
* Closes #4955
* If an item in the spec is a section but the user specified an option, raise an IllegalItemError if it is.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Review by wxtim — COMMENTED** [review]
I think it's worth considering whether the integration test is ok.

File: tests/functional/validate/76-section-as-setting.t [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply tests/functional/validate/76-section-as-setting.t:1]: Turns out that there was special logic to ensure `task parameters` is a section not a setting. This was specific to `task parameters`, this PR is general to all parsec sections so I have removed this specific functionality and its test.

[the remark this answers is not in the document]
Comment (wxtim) [reply tests/functional/validate/76-section-as-setting.t:1]: Happy with what you've done here - seems legit. 👍🏼

**Review by wxtim — APPROVED** [review]
I'm happy with this - 
I think it's big enough to be worth tagging @MetRonnie  or @hjoliver  (Who-ever gets there first) as second review.

**Review by MetRonnie — APPROVED** [review]
Tested out a few examples.

I think some items can be tripped up earlier by the deprecation upgrader. E.g.
```ini
# flow.cylc
scheduling = 22
```
```
  File "~/cylc-flow-8.1/cylc/flow/scripts/validate.py", line 154, in wrapped_main
    cfg = WorkflowConfig(
  File "~/cylc-flow-8.1/cylc/flow/config.py", line 308, in __init__
    self.pcfg = RawWorkflowConfig(
  File "~/cylc-flow-8.1/cylc/flow/cfgspec/workflow.py", line 2042, in __init__
    self.loadcfg(fpath, "workflow definition")
  File "~/cylc-flow-8.1/cylc/flow/parsec/config.py", line 84, in loadcfg
    self.upgrader(sparse, title)
  File "~/cylc-flow-8.1/cylc/flow/cfgspec/workflow.py", line 1891, in upg
    u.upgrade()
  File "~/cylc-flow-8.1/cylc/flow/parsec/upgrade.py", line 200, in upgrade
    old = self.get_item(upg['old'])
  File "~/cylc-flow-8.1/cylc/flow/parsec/upgrade.py", line 107, in get_item
    item = item[key]
TypeError: string indices must be integers
```
Also I noticed the opposite (using a setting as a section) gives traceback too, seems like a low priority though
```ini
# flow.cylc
[scheduling]
    [[cycling mode]]
```
```
  File "~/cylc-flow-8.1/cylc/flow/scripts/validate.py", line 154, in wrapped_main
    cfg = WorkflowConfig(
  File "~/cylc-flow-8.1/cylc/flow/config.py", line 308, in __init__
    self.pcfg = RawWorkflowConfig(
  File "~/cylc-flow-8.1/cylc/flow/cfgspec/workflow.py", line 2042, in __init__
    self.loadcfg(fpath, "workflow definition")
  File "~/cylc-flow-8.1/cylc/flow/parsec/config.py", line 86, in loadcfg
    self.validate(sparse)
  File "~/cylc-flow-8.1/cylc/flow/parsec/config.py", line 96, in validate
    return self.validator(sparse, self.spec)
  File "~/cylc-flow-8.1/cylc/flow/parsec/validate.py", line 1130, in cylc_config_validate
    return CylcConfigValidator().validate(cfg_root, spec_root)
  File "~/cylc-flow-8.1/cylc/flow/parsec/validate.py", line 216, in validate
    cfg[key] = self.coercers[specval.vdr](value, keys + [key])
  File "~/cylc-flow-8.1/cylc/flow/parsec/validate.py", line 389, in coerce_str
    value = cls.strip_and_unquote(keys, value)
  File "~/cylc-flow-8.1/cylc/flow/parsec/validate.py", line 528, in strip_and_unquote
    if value.startswith(substr):
AttributeError: 'OrderedDictWithDefaults' object has no attribute 'startswith'
```

**Comment by oliver-sanders** [comment]
Will haver a quick look to see what can be done for those cases.

**Comment by oliver-sanders** [comment]
@MetRonnie, I've expanded this PR to cover the two additional cases you found.

File: cylc/flow/parsec/validate.py [hunk 2]
```diff
@@ -204,10 +204,28 @@ def validate(self, cfg_root, spec_root):
...
+                cfg_is_section = isinstance(value, dict)
+                spec_is_section = not specval.is_leaf()
+                if cfg_is_section and not spec_is_section:
+                    # config is a [section] but it should be a setting=
+                    raise IllegalItemError(
+                        keys,
+                        key,
+                        msg=f'"{key}" should be a setting not a [section]',
+                    )
+                if not cfg_is_section and spec_is_section:
```
Comment (MetRonnie) [inline cylc/flow/parsec/validate.py:217]: Like when an Oxford comma is missing, my brain doesn't like the ambiguity
```suggestion
                if (not cfg_is_section) and spec_is_section:
```
or
```suggestion
                if spec_is_section and not cfg_is_section:
```

File: cylc/flow/parsec/validate.py [hunk 3]
```diff
@@ -204,10 +204,28 @@ def validate(self, cfg_root, spec_root):
...
+                        keys,
+                        key,
+                        msg=f'"{key}" should be a [section] not a setting',
+                    )
+
+                if cfg_is_section and spec_is_section:
                     # Item is dict, push to queue
                     queue.append([value, specval, keys + [key]])
-                elif value is not None and specval.is_leaf():
+                if value is not None and not spec_is_section:
```
Comment (MetRonnie) [inline cylc/flow/parsec/validate.py:228]: Why the change of `elif` -> `if` here?

File: cylc/flow/parsec/validate.py (hunk 2 above)
Comment (oliver-sanders) [reply cylc/flow/parsec/validate.py]: I'll parenthesise it (because I like brackets), but I wouldn't *expect* anyone to, at least for simple expressions like this. It's Python's bidmas, and & or bind more tightly than not.

git grep 'if not.*and'

**Comment by wxtim** [comment]
Just deconflicted CHANGES.md

File: cylc/flow/parsec/validate.py (hunk 2 above)
Comment (MetRonnie) [reply cylc/flow/parsec/validate.py]: I know but sometimes it's nice to use parentheses in maths too for ease of reading e.g. `6 / 2 * 3` vs `(6 / 2) * 3`
